workspaces/cli-shared/src/captures/avro/file-system/ingest-s3-capture-saver.ts

```diff
@@ -0,0 +1,161 @@
+import { Command } from "commander";
+import { IIgnoreRunnablePredicate } from '@useoptic/cli-config/build/helpers/ignore-parser';
+
+export async function ingestS3({ 
+  bucketName, 
+  region, 
+  captureId, 
+  pathPrefix, 
+  endpointOverride
+}: {
+  bucketName: string,
+  region?: string,
+  captureId: string,
+  pathPrefix?: string,
+  endpointOverride?: string
+}) {
+  let { capturesPath: capturesBaseDirectory, opticIgnorePath, configPath } = await getPathsRelativeToConfig();
+
+  let ignoreRules = await new IgnoreFileHelper(opticIgnorePath, configPath).getCurrentIgnoreRules();
+  let parsedRules = ignoreRules
+    .allRules
+    .map(r => parseRule(r))
+    .filter((r): r is IIgnoreRunnablePredicate => r !== undefined);
+
+  AWS.config.update({ region });
+
+  // Apparently we need to include the bucket here...
+  // Doesn't work if we leave it off and just pass to `listObjectsV2` :s
+  const s3 = new AWS.S3(endpointOverride ? { endpoint: `${endpointOverride}/${bucketName}`, s3BucketEndpoint: true } : undefined);
+
+  const captureSaver = new CaptureSaver({
+    captureBaseDirectory: capturesBaseDirectory,
+    captureId,
+  });
+  await captureSaver.init();
+
+  const prefix = `${path.join(pathPrefix ?? "", captureId ?? "")}`;
+
+  console.log(`Looking in s3://${bucketName}${prefix} in region ${region}, ${endpointOverride ?? ''}`);
+
+  const listAll = async (token?: ContinuationToken): Promise<string[]> => {
+    const objects = await s3.listObjectsV2({
+      Bucket: bucketName,
+      MaxKeys: 1000,
+      Prefix: prefix,
+      ContinuationToken: token
+    }).promise();
+    if (objects.Contents) {
+      let obj_keys = objects.Contents.map(o=>o.Key).filter((o): o is string => !!o);
+      if (objects.IsTruncated) {
+        let rest = await listAll(objects.NextContinuationToken);
+        return obj_keys.concat(rest);
+      } else {
+        return obj_keys;
+      }
+    } else {
+      console.log("No object contents?");
+      return [];
+    }
+  }
+
+  cli.action.start("Loading capture batch list");
+  let allKeys = await listAll();
+  cli.action.stop(`Found ${allKeys.length} capture batches to ingest`);
+  let objectsHandled = 0;
+  let passThrough = new stream.PassThrough();
+
+  const customBar = cli.progress({
+    format: 'Ingesting | {bar} | {value}/{total} batches',
+    barCompleteChar: '\u2588',
+    barIncompleteChar: '\u2591',
+  })
+  customBar.start(allKeys.length,0);
+
+  let fetchPump = async () => {
+    for (const key of allKeys) {
+      let objReadStream = s3.getObject({
+        Bucket: bucketName,
+        Key: key
+      }).createReadStream();
+      await pipeAsync(objReadStream, passThrough);
+      objectsHandled++;
+      customBar.update(objectsHandled);
+    }
+  }
+
+  let interactionCount = 0;
+
+  const oboePromose = new Promise<void>((resolve, reject) => {
+    oboe(passThrough)
+      .node('!', (row: IHttpInteraction) => {
```

Comment:
We're already using [`stream-json`](https://www.npmjs.com/package/stream-json) for streaming json parsing, so lets not introduce oboe as well.

Reply:
Done!

Reply:
FWIW, I wasn't introducing oboe -- it's already in there. But refactored as per our call to use stream-json and be way nicer